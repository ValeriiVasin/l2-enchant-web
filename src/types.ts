export enum Mode {
  Drop,
  Destroy,
}

export type EnchantResult = {
  items: number;
  enchants: number;
  chance: number;
};

export type EnchantResultMode = "items" | "enchants" | "both";

export type FormatEnchantFunction = (index: number) => string;

export type TableRowProps = {
  index: number;
  chance: number;
  result: number;
  formatEnchant?: FormatEnchantFunction;
};
