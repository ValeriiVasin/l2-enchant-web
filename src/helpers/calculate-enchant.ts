import { EnchantResult, Mode } from '../types';

export function calculateEnchant(
  chances: number[],
  mode = Mode.Drop,
): Array<EnchantResult> {
  const coef = 100 / chances[0];
  const items = [mode === Mode.Drop ? 1 : coef];
  const enchants = [coef];

  for (let i = 1; i < chances.length; i++) {
    const chance = chances[i];
    const coef = 100 / chance;

    enchants[i] = enchants[i - 1] * coef + coef;
    items[i] = mode === Mode.Drop ? 1 : items[i - 1] * coef;
  }

  return enchants.map((enchants, i) => ({
    chance: chances[i],
    enchants,
    items: items[i],
  }));
}
