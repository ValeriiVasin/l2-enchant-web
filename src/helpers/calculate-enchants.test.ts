import { test } from '@jest/globals';
import { Mode } from '../types';
import { calculateEnchant } from './calculate-enchant';

const chances = [50, 50, 50];

test('strategy: drop', () => {
  expect(calculateEnchant(chances, Mode.Drop)).toMatchSnapshot();
});
