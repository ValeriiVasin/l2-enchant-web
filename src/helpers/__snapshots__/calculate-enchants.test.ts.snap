// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`strategy: destroy 1`] = `
[
  {
    "chance": 50,
    "enchants": 2,
    "items": 2,
  },
  {
    "chance": 50,
    "enchants": 6,
    "items": 4,
  },
  {
    "chance": 50,
    "enchants": 14,
    "items": 8,
  },
]
`;

exports[`strategy: drop 1`] = `
[
  {
    "chance": 50,
    "enchants": 2,
    "items": 1,
  },
  {
    "chance": 50,
    "enchants": 6,
    "items": 1,
  },
  {
    "chance": 50,
    "enchants": 14,
    "items": 1,
  },
]
`;
